drivers/dht/dht.c

```diff
@@ -67,79 +95,118 @@ int dht_init(dht_t *dev, const dht_params_t *params)
-    }
-
-    /* parse the RAW values */
-    DEBUG("RAW values: temp: %7i hum: %7i\n", (int)raw_temp, (int)raw_hum);
-    switch (dev->type) {
-        case DHT11:
-            *temp = (int16_t)((raw_temp >> 8) * 10);
-            *hum = (int16_t)((raw_hum >> 8) * 10);
-            break;
-        case DHT22:
-            *hum = (int16_t)raw_hum;
-            /* if the high-bit is set, the value is negative */
-            if (raw_temp & 0x8000) {
-                *temp = (int16_t)((raw_temp & ~0x8000) * -1);
+    assert(dev);
+
+    uint32_t now_us = xtimer_now_usec();
+    if ((now_us - dev->last_read_us) > DATA_HOLD_TIME) {
+        /* send init signal to device */
+        gpio_clear(dev->params.pin);
+        xtimer_usleep(20 * US_PER_MS);
+        gpio_set(dev->params.pin);
+        xtimer_usleep(40);
+
+        /* sync on device */
+        gpio_init(dev->params.pin, dev->params.in_mode);
+        unsigned counter = 0;
+        while (!gpio_read(dev->params.pin)) {
+            if (counter++ > TIMEOUT) {
+                _reset(dev);
+                return DHT_TIMEOUT;
             }
-            else {
-                *temp = (int16_t)raw_temp;
+            xtimer_usleep(1);
+        }
+
+        counter = 0;
+        while (gpio_read(dev->params.pin)) {
+            if (counter++ > TIMEOUT) {
+                _reset(dev);
+                return DHT_TIMEOUT;
             }
-            break;
-        default:
-            return DHT_NODEV;      /* this should never be reached */
+            xtimer_usleep(1);
+        }
+
+        /*
+         * data is read in sequentially, highest bit first:
+         *  40 .. 24  23   ..   8  7  ..  0
+         * [humidity][temperature][checksum]
+         */
+
+        /* read the humidity, temperature, and checksum bits */
+        if (_read(&raw_hum, dev->params.pin, 16)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        if (_read(&raw_temp, dev->params.pin, 16)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        if (_read(&csum, dev->params.pin, 8)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        /* Bring device back to defined state - so we can trigger the next reading
+         * by pulling the data pin low again */
+        _reset(dev);
+
+        /* validate the checksum */
+        sum = (raw_temp >> 8) + (raw_temp & 0xff) + (raw_hum >> 8) + (raw_hum & 0xff);
+        if ((sum != csum) || (csum == 0)) {
+            DEBUG("error: checksum invalid\n");
+            return DHT_NOCSUM;
+        }
+
+        /* parse the RAW values */
+        DEBUG("RAW values: temp: %7i hum: %7i\n", (int)raw_temp, (int)raw_hum);
+        switch (dev->params.type) {
+            case DHT11:
+                dev->last_val.temperature = (int16_t)((raw_temp >> 8) * 10);
+                dev->last_val.humidity = (int16_t)((raw_hum >> 8) * 10);
+                break;
+            case DHT22:
```

Comment:
where is `DHT21` which is also allowed according to the assert in init

Reply:
Both DHT21 and DHT22 have the same value in the enum. If I remember correctly the compiler does not like using the same value twice in a `switch`, even if they fall through.

(Actually, I only adjusted the indent here and did not touch the code here. But I can add a comment to explain the reason of the original author to omit DHT21.)

Reply:
yeah sorry, just saw it wasn't there before. ignore then